src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Window.cs

```diff
@@ -4710,11 +4710,7 @@ internal virtual void WmMoveChangedHelper()
 
         private bool WmGetMinMaxInfo( IntPtr lParam )
         {
-            NativeMethods.MINMAXINFO mmi;
```

Comment:
Hey @stephentoub, could I have your opinion here please?

In one of your PR, you changed Marshal.PtrToStructure for this unsafe code: https://github.com/dotnet/wpf/pull/4754/files#diff-30ce94d35c72b94385aa439cafead5bd0677f7b4037f0fefcfc7a03d3730a681L4702-R4705

What is best, the unsafe code or generic Marshal.PtrToStructure ?

If it's the unsafe code, when should we use it ? Should I change my PR to convert Marshal.PtrToStructure to this unsafe code ?

Thank you!

Reply:
Both are equally "unsafe", so it's purely a performance / functionality discussion.  If the type isn't blittable, use `Marshal.PtrToStructure`, as it appropriately handles all the necessary marshaling.  If the type is blittable, it's _way_ cheaper to just cast rather than using `PtrToStructure`.

|         Method |       Mean | Allocated |
|--------------- |-----------:|----------:|
|           Cast |  0.0821 ns |         - |
| PtrToStructure | 25.5936 ns |      24 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Runtime.InteropServices;

[MemoryDiagnoser]
public class Program
{
    public static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private IntPtr _ptr;

    [GlobalSetup]
    public unsafe void Setup() => _ptr = Marshal.AllocHGlobal(sizeof(MyPoint));

    [GlobalCleanup]
    public void Cleanup() => Marshal.FreeHGlobal(_ptr);

    [Benchmark]
    public unsafe MyPoint Cast() => *(MyPoint*)_ptr;

    [Benchmark]
    public MyPoint PtrToStructure() => Marshal.PtrToStructure<MyPoint>(_ptr);
}

[StructLayout(LayoutKind.Sequential)]
public struct MyPoint
{
    public int X;
    public int Y;
}
```

Reply:
Thank you!

> Both are equally "unsafe", so it's purely a performance / functionality discussion
>
I didn't mean "unsafe" as "this code is less safe than this one", I just didn't know how to name your code and I used "unsafe" because it used unsafe blocks.

I'll revert this change and keep the usage of generic PtrToStructure everywhere else in this PR since the generic version is slightly faster than the non-generic.

I'll open an issue to use your code to cast IntPtr to blittable structs (Maybe also add a shared method to avoid having unsafe blocks everywhere to improve readability).

Reply:
@stephentoub Can I ask you why the PtrToStructureGeneric is faster than PtrToStructureNonGeneric?

See:

https://github.com/dotnet/runtime/blob/52c6dcb45cc1e8e98d267c07f10fe42d89c71656/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L575-L600

https://github.com/dotnet/runtime/blob/52c6dcb45cc1e8e98d267c07f10fe42d89c71656/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L615-L634

Thank you.

Reply:
For the generic, and a value type, the JIT generates a specialization of the method for that value type, such that the PtrToStructure method knows what the type is, which means it can do at compile time things the non-generic needs to do at runtime.

Reply:
@stephentoub  Thank you very much.